File: Code/Mantid/PythonAPI/PythonAlgorithms/ExtractFFTSpectrum.py
from MantidFramework import *
from mantidsimple import *

class ExtractFFTSpectrum(PythonAlgorithm):

	def category(self):
		return "MSG"
	def name(self):
		return "ExtractFFTSpectrum"

	def PyInit(self):
		self.declareWorkspaceProperty("InputWorkspace", "", Direction.Input)
		self.declareWorkspaceProperty("OutputWorkspace", "", Direction.Output)
		self.declareProperty("FFTPart", 2)
		
	def PyExec(self):
		inWS = self.getProperty("InputWorkspace")
		fftPart = self.getProperty("FFTPart")
		nhist = inWS.getNumberHistograms()
		nbins = inWS.getNumberBins()
		
		for i in range(0, nhist):
			tmpWS = '_fury_algorithm_temp'
			FFT(inWS, tmpWS, i)
			
			tempWS = mantid.getMatrixWorkspace(tmpWS)
			if ( i == 0 ):
				outWS = WorkspaceFactory.createMatrixWorkspaceFromCopy(tempWS, nhist, nbins+1, nbins)
			for j in range(0, outWS.getNumberBins()):
				outWS.dataX(i)[j] = tempWS.readX(fftPart)[j]
				outWS.dataE(i)[j] = tempWS.readE(fftPart)[j]
				outWS.dataY(i)[j] = tempWS.readY(fftPart)[j]
			outWS.dataX(i)[nbins] = tempWS.readX(fftPart)[nbins]
		
		mantid.deleteWorkspace(tmpWS)
		self.setProperty("OutputWorkspace", outWS)

mantid.registerPyAlgorithm(ExtractFFTSpectrum())

File: Code/Mantid/Algorithms/src/ExtractFFTSpectrum.cpp
//----------------------------------------------------------------------
// Includes
//----------------------------------------------------------------------
#include "MantidAlgorithms/ExtractFFTSpectrum.h"
#include "MantidKernel/MultiThreaded.h"
#include "MantidKernel/UnitFactory.h"

namespace Mantid
{
namespace Algorithms
{

// Register the algorithm into the AlgorithmFactory
DECLARE_ALGORITHM(ExtractFFTSpectrum)

using namespace Kernel;
using namespace API;

void ExtractFFTSpectrum::init()
{
  declareProperty(new WorkspaceProperty<>("InputWorkspace","",Direction::Input),"The input workspace.");
  declareProperty("FFTPart", 2, new BoundedValidator<int>(0,5));
  declareProperty(new WorkspaceProperty<>("OutputWorkspace","",Direction::Output), "The output workspace.");
}

void ExtractFFTSpectrum::exec()
{
  MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
  const int fftPart = getProperty("FFTPart");
  const int numHists = inputWS->getNumberHistograms();
  MatrixWorkspace_sptr outputWS = WorkspaceFactory::Instance().create(inputWS);

  Progress prog(this, 0.0, 1.0, numHists);

  PARALLEL_FOR1(outputWS)
  for ( int i = 0; i < numHists; i++ )
  {
    PARALLEL_START_INTERUPT_REGION

    IAlgorithm_sptr childFFT = createSubAlgorithm("FFT");
    childFFT->setProperty<MatrixWorkspace_sptr>("InputWorkspace", inputWS);
    childFFT->setProperty<int>("Real", i);
    childFFT->execute();
    MatrixWorkspace_sptr fftTemp = childFFT->getProperty("OutputWorkspace");

    outputWS->dataE(i) = fftTemp->readE(fftPart);
    outputWS->dataY(i) = fftTemp->readY(fftPart);
    outputWS->dataX(i) = fftTemp->readX(fftPart);

    prog.report();

    PARALLEL_END_INTERUPT_REGION
  }
  PARALLEL_CHECK_INTERUPT_REGION

  boost::shared_ptr<Kernel::Units::Label> lblUnit = boost::dynamic_pointer_cast<Kernel::Units::Label>(UnitFactory::Instance().create("Label"));
  lblUnit->setLabel("Time", "ns");
  outputWS->getAxis(0)->unit() = lblUnit;

  setProperty("OutputWorkspace", outputWS);
}
} // namespace Algorithms
} // namespace Mantid
